### PR TITLE
VDS: properly handle the clone cache key variant during client callback execution

### PR DESCRIPTION
**Pull request by benashz**
In the case where a VaultDynamicSecret instance specifies a Vault namespace that is different from its VaultAuth the ClientCacheKey will be of the clone variant. The cache key format is different from its parent cache key and therefore a parent and a clone cannot be compared. Since the VDS controller stores the clone variant in its state, all instances of this type were ignored during a client callback reconciliation. This would leave unreconciled VDS instances upon client token expiration or other LifetimeWatcher errors.

- [x] Extend the VDS integration tests to include a cross-namespace scenario. (disabled until https://github.com/hashicorp/terraform-provider-vault/pull/2289 is released)

### CODE REVIEW

File: controllers/vaultdynamicsecret_controller.go [hunk 1]
```diff
@@ -790,7 +790,14 @@ func (r *VaultDynamicSecretReconciler) vaultClientCallback(ctx context.Context,
 
 	reqs := map[client.ObjectKey]empty{}
 	for _, o := range l.Items {
-		if o.Status.VaultClientMeta.CacheKey == cacheKey.String() {
+		if o.Status.VaultClientMeta.CacheKey == "" {
+			logger.V(consts.LogLevelWarning).Info("Skipping, cacheKey is empty",
+				client.ObjectKeyFromObject(&o))
```
Comment (tvoran) [inline controllers/vaultdynamicsecret_controller.go:795]:
```suggestion
			logger.V(consts.LogLevelWarning).Info("Skipping, cacheKey is empty",
				"object", client.ObjectKeyFromObject(&o))
```

**Review by tvoran — APPROVED** [review]

File: internal/vault/cache_key_test.go [hunk 2]
```diff
@@ -337,3 +338,125 @@ func TestClientCacheKeyClone(t *testing.T) {
...
+			want:    false,
+			wantErr: assert.Error,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.key.SameParent(tt.other)
+			if !tt.wantErr(t, err, fmt.Sprintf("SameParent(%v)", tt.other)) {
+				return
+			}
```
[the remark this answers is not in the document]
Comment (benashz) [reply internal/vault/cache_key_test.go]: Fixed in 53c5d9fc55f47fd9f9acdc113957617ddaaa4b12

File: controllers/vaultdynamicsecret_controller.go (hunk 1 above)
Comment (benashz) [reply controllers/vaultdynamicsecret_controller.go]: Fixed in 53c5d9fc55f47fd9f9acdc113957617ddaaa4b12

File: test/integration/vaultdynamicsecret_integration_test.go [hunk 3]
```diff
@@ -826,6 +851,14 @@ func TestVaultDynamicSecret_vaultClientCallback(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			if tt.xns {
+				t.Skipf("skipping xns test %s, until https://github.com/hashicorp/terraform-provider-vault/pull/2289 is released", tt.name)
+			}
+
+			if tt.xns && !outputs.WithXns {
+				t.Skipf("skipping xns test %s, test infrastructre not supported", tt.name)
```
[the remark this answers is not in the document]
Comment (benashz) [reply test/integration/vaultdynamicsecret_integration_test.go]: Fixed in 53c5d9fc55f47fd9f9acdc113957617ddaaa4b12

**Pull request merged**